# Data Science Masters Challenge
## Assignment -4:

<b> Q1. </b> What exactly is [ ]?

<b>Answer: </b>

It indicates the list. It is creating an empty list.

In [1]:
type([])

list

list

<b> Q2. </b> In a list of values stored in a variable called spam, how would you assign the value 'hello' as the
third value? (Assume [2, 4, 6, 8, 10] are in spam.)

<b> Answer: </b>

In [2]:
spam = [2,4,6,8,10]
print(spam)
spam[2] = "hello"
print(spam)

[2, 4, 6, 8, 10]
[2, 4, 'hello', 8, 10]


[2, 4, 6, 8, 10]
<br>[2, 4, 'hello', 8, 10]

Let's pretend the spam includes the list ['a', 'b', 'c', 'd'] for this queries.

<b>Q3. </b>What is the value of spam[int(int('3'* 2) / 11)]?

<b> Answer:</b>

The value of the above code is : <b> 'd' </b>

In [3]:
spam = ['a', 'b', 'c', 'd'] 
spam[int(int('3'* 2) / 11)]# = spam[int(33/11)]

'd'

Let's pretend the spam includes the list ['a', 'b', 'c', 'd'] for this queries.

<b> Q4. </b>What is the value of spam[-1]?

<b> Answer : </b>

d

In [4]:
spam = ['a', 'b', 'c', 'd'] 
spam[-1]

'd'

Let's pretend the spam includes the list ['a', 'b', 'c', 'd'] for this queries.

<b> Q5. </b>What is the value of spam[:2]?

<b> Answer : </b>

['a', 'b']

In [5]:
spam = ['a', 'b', 'c', 'd'] 
spam[:2]

['a', 'b']

Let's pretend bacon has the list [3.14, 'cat',11, 'cat',True] for this questions.

<b>Q6. </b>What is the value of bacon.index('cat')?

<b> Answer: </b>

1

In [6]:
bacon = [3.14, 'cat',11, 'cat',True]
bacon.index('cat')

1

Let's pretend bacon has the list [3.14, 'cat',11, 'cat',True] for this questions.

<b>Q7. </b>How does bacon.append(99) change the look of the list value in bacon?

<b>Answer: </b>

[3.14, 'cat', 11, 'cat', True]
<br>[3.14, 'cat', 11, 'cat', True, 99]

In [7]:
bacon = [3.14, 'cat',11, 'cat',True]
print(bacon)
bacon.append(99) # it will add 99 to end of the list 
print(bacon)

[3.14, 'cat', 11, 'cat', True]
[3.14, 'cat', 11, 'cat', True, 99]


Let's pretend bacon has the list [3.14, 'cat',11, 'cat',True] for this questions.

<b>Q8. </b>How does bacon.remove('cat') change the look of the list in bacon?

<b> Answer: </b>

[3.14, 'cat', 11, 'cat', True]
<br>[3.14, 11, 'cat', True]

In [8]:
bacon = [3.14, 'cat',11, 'cat',True]
print(bacon)
bacon.remove('cat') # remove the element in the list from the location it has appeared for the first time.
print(bacon)

[3.14, 'cat', 11, 'cat', True]
[3.14, 11, 'cat', True]


<b> Q9. </b> What are the list concatenation and list replication operators?

<b> Answer: </b>

In [9]:
# list concatination operator
l1 =["a",2]
l2 = [5.6,4+6j]
print("list concatination operation : ",l1 + l2)

# list replication operator
l1 =["a",2]
print("list replication operation   : ",l1*2)

list concatination operation :  ['a', 2, 5.6, (4+6j)]
list replication operation   :  ['a', 2, 'a', 2]


list concatination operation :  ['a', 2, 5.6, (4+6j)]
<br>list replication operation   :  ['a', 2, 'a', 2]

<b>Q10. </b> What is difference between the list methods append() and insert()?

<b>Answer: </b>

In [10]:
l1 =["a",2]
print("Original list    : ",l1)
l1.append("v")
print("append operation : ",l1) # add the element at the end of the list
l1.insert(1,56.7)
print("insert opeartion : ",l1) # can add element to any index location

Original list    :  ['a', 2]
append operation :  ['a', 2, 'v']
insert opeartion :  ['a', 56.7, 2, 'v']


Original list    :  ['a', 2]
<br>append operation :  ['a', 2, 'v']
<br>insert opeartion :  ['a', 56.7, 2, 'v']

<b>Q11. </b> What are the two methods for removing items from a list?

<b> Answer: </b>

In [11]:
l = ['a', 56.7, 2, 'v',8+6j,"verify",True]
print("Original list       : ",l)
l.remove(8+6j)
print("After remove method : ",l)
l.pop(1)
print("After pop method    : ",l)

# we also can discard the element using del()
del(l[3])
print("After del()         : ",l)

Original list       :  ['a', 56.7, 2, 'v', (8+6j), 'verify', True]
After remove method :  ['a', 56.7, 2, 'v', 'verify', True]
After pop method    :  ['a', 2, 'v', 'verify', True]
After del()         :  ['a', 2, 'v', True]


Original list       :  ['a', 56.7, 2, 'v', (8+6j), 'verify', True]
<br>After remove method :  ['a', 56.7, 2, 'v', 'verify', True]
<br>After pop method    :  ['a', 2, 'v', 'verify', True]
<br>After del( )         :  ['a', 2, 'v', True]

<b>Q12. </b> Describe how list values and string values are identical.

<b>Answer: </b>

Lists are similar to strings, which are ordered collections of characters, except that the elements of a list can have any type and for any one list, the items can be of different types.
<li>Strings can be defined as sequential collections of characters and a list is a sequential collection of Python data values, where each value is identified by an index.</li>
<li> String is a ordered collection of character, List is also a ordered collection, where it contains elements of any type. </li>
<li> for any list, the items in it can be of different types.</li>

<b>Q13. </b>What's the difference between tuples and lists?

<b> Answer : </b> 
<table style="border:1px solid black;" align="left">
    <tr><th>Tuples</th> <th>lists</th> </tr>
    <tr><td> immutable </td> <td>mutable</td></tr>
    <tr><td> elements are defined within ( ) </td> <td>elements are defined within [ ] </td></tr>
    <tr><td> e.g. p = (1,"a",3.4,True,5+7j,(4,5),["a",5],{1,2,"s"},{1:"a"}) </td><td>e.g. p = [1,"a",3.4,True,5+7j,(4,5),["a",5],{1,2,"s"},{1:"a"}]</td></tr>
    <tr><td> The methods supported under tuples are : <li>count</li><li> index</li> </td> <td>The methods supported under tuples are : <li>append </li><li>clear</li><li>copy</li> <li>count</li> <li>extend</li><li>index</li><li>insert</li><li>pop</li><li>remove</li><li>reverse</li><li>sort</li> </td></tr>
</table>

<b>Q14. </b>How do you type a tuple value that only contains the integer 42?

<b> Answer : </b>

In [12]:
t = (1,"a",3.4,True,5+7j,(4,5),42,["a",5],{1,2,"s"},{1:"a"})
for k,v in enumerate(t):
    if v == 42:
        print(" index :",k,"value :",v)

 index : 6 value : 42


index : 6 value : 42

<b>Q15.How do you get a list value's tuple form? How do you get a tuple value's list form?

<b> Answer : </b>

In [13]:
t = tuple()
l = list()
print("Type of l =",type(l),"Type of t =",type(t))
t = list(t)
l = tuple(l)
print("After type casting")
print("Type of l =",type(l),"Type of t =",type(t))

Type of l = <class 'list'> Type of t = <class 'tuple'>
After type casting
Type of l = <class 'tuple'> Type of t = <class 'list'>


Type of l = <class 'list'> Type of t = <class 'tuple'>
<br>After type casting
<br>Type of l = <class 'tuple'> Type of t = <class 'list'>

<b>Q16. </b> Variables that "contain" list values are not necessarily lists themselves. Instead, what do they
contain?

<b> Answer: </b> 
    
Python variable is a symbolic name that is a reference or pointer to an object.

<b>Q17. </b> How do you distinguish between copy.copy() and copy.deepcopy()?

 <b> Answer: </b>
 
<table style="border:1px solid black;" align="left">
    <tr><th>Shallow copy</th> <th>Deep copy</th> </tr>
    <tr><td> <b>Denoted as :</b> copy.copy( ) </td> <td><b>Denoted as :</b> copy.deepcopy( )</td></tr>
    <tr><td>The copying process does not recurse and therefore won’t create copies of the child objects  </td> <td>Deep copy is a process in which the copying process occurs recursively.</td></tr>
    <tr><td>It first construct a new collection object and then populating it with references to the child objects found in the original.  </td><td>First it construct a new collection object and then recursively populating it with copies of the child objects found in the original.</td></tr>
    <tr><td>A reference of object is copied in other object</td><td>A copy of object is copied in other object</td></tr>
    <tr><td><b><u> Program:</u></b><br>import copy<br>l1 = [1, 2, [3,4], 5]<br>print("l1 =",l1)<br>l2 = copy.deepcopy(l1) <br>l2[2][0] = 0<br>print("After copy.deepcopy()")<br>print("l1 =",l1)<br>print("l2 =",l2)<br><b><u>Output:</u></b><br>l1 = [1, 2, [3, 4], 5]
<br>After copy.copy()
<br>l1 = [1, 2, [0, 4], 5]
<br>l2 = [1, 2, [0, 4], 5]</td><td><b><u> Program:</u></b><br>import copy<br>l1 = [1, 2, [3,4], 5]<br>print("l1 =",l1)<br>l2 = copy.deepcopy(l1) <br>l2[2][0] = 0<br>print("After copy.deepcopy()")<br>print("l1 =",l1)<br>print("l2 =",l2)<br><b><u>Output:</u></b><br>l1 = [1, 2, [3, 4], 5]
<br>After copy.deepcopy()
<br>l1 = [1, 2, [3, 4], 5]
<br>l2 = [1, 2, [0, 4], 5]</td></tr>
</table>

In [14]:
import copy
l1 = [1, 2, [3,4], 5]
l2 = copy.copy(l1) 
print("Original values","      l1 =",l1,"                       id(l1) :",id(l1))
l2[2][0] = 0
print("With copy.copy()","     l1 =",l1,"l2 =",l2,"id(l1) :",id(l1),"ld(l2) :",id(l2))

l2 = copy.deepcopy(l1)
l2[2][0] = 8
print("With copy.deepcopy()"," l1 =",l1,"l2 =",l2,"id(l1) :",id(l1),"id(l2) :",id(l2))

Original values       l1 = [1, 2, [3, 4], 5]                        id(l1) : 2812472559168
With copy.copy()      l1 = [1, 2, [0, 4], 5] l2 = [1, 2, [0, 4], 5] id(l1) : 2812472559168 ld(l2) : 2812472561152
With copy.deepcopy()  l1 = [1, 2, [0, 4], 5] l2 = [1, 2, [8, 4], 5] id(l1) : 2812472559168 id(l2) : 2812472847616


Original values       l1 = [1, 2, [3, 4], 5]                        id(l1) : 2812472559168
<br>With copy.copy()      l1 = [1, 2, [0, 4], 5] l2 = [1, 2, [0, 4], 5] id(l1) : 2812472559168 ld(l2) : 2812472561152
<br>With copy.deepcopy()  l1 = [1, 2, [0, 4], 5] l2 = [1, 2, [8, 4], 5] id(l1) : 2812472559168 id(l2) : 2812472847616